### ISSUE TITLE
add labels to important GCP resources

### ISSUE THREAD

**Issue by shaneknapp** [berkeley-dsep-infra/hubs#4381]
this will be an ongoing project, and any policy/process that we refine here will need to be documented fully.

TL;DR:  we need to label (not tag) all resources in GCP.  this will allow us, at the minimum to have much more granular billing data.  in a broader context, this will make searching for all labeled resources much easier (web console and CLI).  labels also help out w/BigQuery if (when, really) we start using that service.

the docs are pretty terrible, and i wasted a bunch of time confirming that only some things (like filestore instances) allow you to add/edit/delete labels both in the web console and CLI, but some things (node-pools, i'm looking at you) have no web console support for labels but full support in the CLI. 

https://cloud.google.com/resource-manager/docs/creating-managing-labels
https://cloud.google.com/kubernetes-engine/docs/how-to/creating-managing-labels#labels_and_tags

another confusing thing is that labels are not tags and particularly, not either metadata or kubernetes labels.  yes:  labels are not kubernetes labels, even though they're labels.

just look at the man pages for things like ` gcloud filestore instances create` and `gcloud container node-pools update` and search for `labels`.  the thing we're most interested in is `--labels`.  :)

anyways.  these are the four main things that need to be completed:

## determining naming conventions and data models for all labels

since these are key/value pairs, let's pick some good keys and value!  this area will have the biggest end-user (us) impact, so we'll need to think carefully about how we go about implementing things.

off of the top of my head:
  - have a `deployment:<hub-name>` label on all top-level resources that a hub needs that allows labels.  filestore, node-pools, backups, whatever.  this should allow for lots more billing and usage data on a course-by-course basis.
  - prod, staging, ???

we should also try not to over-label everything and do our best to remove any old/stale labels.

## update existing resources w/labels

pretty straightforward, tbh.
  - [x] figure out what we want to label
  - [x] figure out what those labels will be
  - [x] apply labels

## update documentation for new hub creation, general doc for labeling

  - [x] make sure that labels are properly applied when new stuff is deployed
  - [x] document our labeling schema

## create some billing reports and figure out how much courses cost!

should be pretty easy, i believe.

i'll start experimenting by adding some labels to data100 and mucking about.  :)

**Comment by shaneknapp** [comment]
some notes:
1.  i already (manually) created some filestore labels (`filestore-deployment=hubname`).  these may have limited usefulness and may need to be renamed
2. it took a large number of hours for the labels i created in filestore to propagate through the series of tubes and become available in the billing console.  basically "next day" service...  will see if this is observed when updating things through the CLI (see next item)
3. i'm manually added a `hub=data100` label to both filestore and GKE.  it's taking for-ev-er for `gcloud container node-pools update user-data100-2022-10-28 --project "ucb-datahub-2018" --cluster "fall-2019" --region "us-central1" --labels hub=data100` to finish running.

**Comment by ryanlovett** [comment]
This all sounds great! Agree strongly about ensuring things are labeled when they're deployed and documenting everything. I do like the notion of deployment={deployment_name} and hub={hub_name} tags. I guess the latter would be for when we need to track resources down to prod and staging.

**Comment by shaneknapp** [comment]
updates/observations:
1.  confirmed that the labels take ~overnight to propagate and become visible in the billing console
2. AFAICT we will not be able to break out individual hub costs that use shared node pools and/or filestore instances
3. label propagation is a little bit wonky -- even though i labeling the data100 node pool, i can't search for `hub:data100` in `GKE -> clusters -> node pools` or `GKE -> clusters -> nodes`.  *however*, since the nodes are VMs, you CAN find them by searching for label/value in `compute engine -> VM instances`.  the *only* other place i could find it was in annoying.
4. in fact, the label rabbit hole continues.  of course it does.  on the CLI, you can find the node pool labels using `gcloud`.  the labels are nested in a `resourceLabel` namespace.  all other labels were in, you guessed it, `labels`.  so node pool labels are `resourceLabels` and seem to be applied to all 'resources' used by the node pool.  see next item in this list after reading [this](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/ResourceLabels) and assuming that `resourceLabels` are `labels` given that's what the code implies?  i didn't look too deep
5. Here's how it works:  node pool labels are applied to the node pool **instance group**, and then any members of that group will have the label applied to them.  you can only see this label in the web console by traversing through the instance group, then clicking on the instance name.  if you click directly on the instance name in the node pool details page, it will not display the label, export it, or be visible in the config.
6. we still need to identify which GCP services are appropriate to label for each hub.  i currently only have Compute Engine and filestore labeled, but there are 12 additional services (of which most probably don't matter) to think about.  CE and filestore will by FAR be the biggest source of spends but it'd be nice to know exactly what we should track and what it costs.

some screenies:

view of a per-hub daily spend (right now just data100 CE + filestore) [source link](https://console.cloud.google.com/billing/013554-935B0A-B97AA1/reports;timeRange=CUSTOM_RANGE;from=2023-03-20;to=2023-03-21;dateType=USAGE_DATE;products=ALL;labels=hub:data100;savedView=576fca29-4364-49c1-bcba-e076bd6f37e0?organizationId=70968413100&project=ucb-datahub-2018):
![Screenshot 2023-03-21 at 9 28 46 AM](https://user-images.githubusercontent.com/1606572/226676334-03037489-c0b5-4e91-afba-8eea79427a67.png)

searching for labeled nodes in the compute engine console:
![Screenshot 2023-03-21 at 9 16 03 AM](https://user-images.githubusercontent.com/1606572/226676349-32c755c0-6aab-419e-b67b-873f5ff0e555.png)

the label in the billing console:
![Screenshot 2023-03-21 at 8 55 15 AM](https://user-images.githubusercontent.com/1606572/226676352-c3f0f745-e097-4ec3-94b1-a45b21702ae9.png)

searching by label in filestore:
![Screenshot 2023-03-21 at 9 34 24 AM](https://user-images.githubusercontent.com/1606572/226678202-2ffa7e16-9f34-4e8c-84cd-29b913c23d49.png)

**Comment by shaneknapp** [comment]
> This all sounds great! Agree strongly about ensuring things are labeled when they're deployed and documenting everything. I do like the notion of deployment={deployment_name} and hub={hub_name} tags. I guess the latter would be for when we need to track resources down to prod and staging.

so maybe something like:
`deployment=hubname`
`hub=<hubname>-(prod|staging)`

it might be hard to label prod vs staging, but i'm just saying that w/o any real knowledge rn.  :)

**Comment by shaneknapp** [comment]
the vernacular:

data definitions:
`name`:  the GCP name for a resource.  this is the first field returned from an invocation of `gcloud filestore instances list ...` or `gcloud container node-pools list ...`
`hub`:  short form of the above.  `support` from the `support-2022-12-08` node pool.  `datahub` from the `datahub-12-20-2022` filestore instance.

the labels:
`hub`: short form of the node pool/filestore instance/filestore backup name.  applied to node pools, filestore and filestore backups.  allows us to see total cost per hub.  caveat:  can't see individual usage for hubs hosted on shared node pools or filestore instances (`small-courses`) 
`nodepool-deployment`:  short form name.  what hub is using this node pool.  applied only to node pools
`filestore-deployment`: short form name.  what hub is using this filestore.  applied only to filestore instances and backups

ok, i plan on executing this, after a bit more review:

```
# update node pools with nodepool-deployment labels
for x in $(cat nodepool-hub-labels.txt); do
  IFS=" " read name hub <<< "$x";
  # $hub is name of the deployment, $name is node pool name
  gcloud container node-pools update $name \
    --project "ucb-datahub-2018" --cluster "fall-2019" \
    --region "us-central1" --labels=nodepool-deployment=$hub ; 
  gcloud container node-pools update $name \
    --project "ucb-datahub-2018" --cluster "fall-2019" \
    --region "us-central1" --labels=hub=$deployment ; 
  ; done

# update filestore instances and backups with hub labels
for x in $(cat filestore-hub-labels.txt); do
  IFS=" " read name hub <<< "$x";
  # $hub is name of the deployment, $name filestore instance name 
  # filestore-deployment label already exists
  gcloud filestore instances update $name \
    --project "ucb-datahub-2018" --cluster "fall-2019" \
    --zone "us-central1-b" --update-labels=hub=$hub ; 
  ; done

for x in $(cat backup-hub-labels.txt); do
  IFS=" " read name hub <<< "$x";
  # hub == hub, backups
  gcloud filestore backups update $name \
    --project "ucb-datahub-2018" --cluster "fall-2019" \
    --zone "us-central1-b" --update-labels=hub=$hub ; 
  gcloud filestore backups update $name \
    --project "ucb-datahub-2018" --cluster "fall-2019" \
    --zone "us-central1-b" --update-labels=filestore-deployment=$hub ; 
  ; done
```
[filestore-hub-labels.txt](https://github.com/berkeley-dsep-infra/datahub/files/11045296/filestore-hub-labels.txt)
[backup-hub-labels.txt](https://github.com/berkeley-dsep-infra/datahub/files/11045301/backup-hub-labels.txt)
[nodepool-hub-labels.txt](https://github.com/berkeley-dsep-infra/datahub/files/11045306/nodepool-hub-labels.txt)

all the names for filestore, backups and node pools was collected this way:
```
gcloud container node-pools list --project "ucb-datahub-2018" --cluster "fall-2019" --region "us-central1" | awk '{print$1}'|grep -v NAME > nodepools
gcloud filestore instances list|awk '{print$1}' | grep -v NAME > filestore-instances
gcloud filestore backups list| grep -v NAME | awk '{print$1}' > backup-names
```

i've attached the files as well...  these are big changes w/lots of impact so i want to make sure i didn't miss anything.  :)

**Comment by balajialg** [comment]
This is awesome work @shaneknapp !
